### DISCUSSION TITLE
Different results for declaration vs function argument

### DISCUSSION THREAD

**Discussion by andrew222651** [microsoft/pyright#7993] [Q&A] [answered]
I'm wondering why this has an error
```python
Array__int4 = list[int | None] | list["Array__int4"] | None

a: Array__int4 = [1,2,[3],None]
```
but this doesn't
```python
Array__int4 = list[int | None] | list["Array__int4"] | None

a = [1,2,[3],None]

def f(arr: Array__int4):
    pass

f(a)
```
In mypy they both have an error.

ETA: If we have `a = [1,2,[3],None]`, then Pyright infers the type as `list[Unknown]` whereas mypy infers it as `list[object]`.

**Comment by erictraut** [comment, marked as answer]
This has to do with pyright's default behavior for list expression type inference. In your second example, the inferred type of `[1, 2, [3], None]` is `list[Unknown]` (where `Unknown` is pyright's way of indicating an "implied `Any`"). For details, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions).

Inference rules are not specified in the typing spec, and pyright's behaviors differ in some ways from mypy's. For an explanation of why pyright's behaviors intentionally differs from mypy in some areas, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=differences-between-pyright-and-mypy).

**Reply by andrew222651** [reply]
Thank you!
